interpose.h

```diff
@@ -0,0 +1,41 @@
+#ifndef __INTERPOSE_H
+#define __INTERPOSE_H
+
+#include <stdint.h>
+#include <dlfcn.h>
+
+#if defined(__APPLE__)
+
+/// Structure exposed to the linker for interposition
+struct __osx_interpose {
+    const void* new_func;
+    const void* orig_func;
+};
+
+/**
+ * Generate a macOS interpose struct
+ * Types from: http://opensource.apple.com/source/dyld/dyld-210.2.3/include/mach-o/dyld-interposing.h
+ */
+#define OSX_INTERPOSE_STRUCT(NEW, OLD) \
```

Comment:
Normally, pre-processor macros would be the right way to eliminate duplicated code.

In this particular case, the C code is generated from Python in `generator/gen_libc_hooks.py`, and there isn't a good way to emit code that calls the C preprocessor macros. It would be better to implement the logic directly in Python.

Therefore, for the above file, can you write out what it should be after expanding the pre-processor macros?

Reply:
My concern is that in Linux, the shared library needs to define a function with the name we want to interpose (e.g., `open`), and load the real/underlying function with a new name (e.g., ` unwrapped_open`) using `dlsym`.

In Mac, it seems that we define a function with the new name (e.g., `my_open`), and call the real/underlying function with the original name (`open`)

```
__attribute__ ((section ("__DATA,__interpose"))) = {
    (const void*)(unsigned long)&my_open,
    (const void*)(unsigned long)&open
};
```

The meaning of the name `open` is reversed: in Linux its the wrapper; in Mac it's the underlying function.

Reply:
This is of course salvageable. We should always call `unwrapped_open` and `wrapped_open` to eliminate confusion, but alias one of them to the "original name" based on platform.

```
#ifdef linux
#define wrapped_open open
#else if apple
#define unwrapped_open open
#endif
```

We hardly ever call the wrapped versions, always the unwrapped, which are already called as `unwrapped_` in our code.